django/db/backends/sqlite3/_functions.py

```diff
@@ -0,0 +1,471 @@
+        return None
+    conn = conn.strip()
+    try:
+        real_lhs = _sqlite_prepare_dtdelta_param(conn, lhs)
+        real_rhs = _sqlite_prepare_dtdelta_param(conn, rhs)
+    except (ValueError, TypeError):
+        return None
+    if conn == '+':
+        # typecast_timestamp returns a date or a datetime without timezone.
+        # It will be formatted as "%Y-%m-%d" or "%Y-%m-%d %H:%M:%S[.%f]"
+        out = str(real_lhs + real_rhs)
+    elif conn == '-':
+        out = str(real_lhs - real_rhs)
+    elif conn == '*':
+        out = real_lhs * real_rhs
+    else:
+        out = real_lhs / real_rhs
+    return out
+
+
+def _sqlite_time_diff(lhs, rhs):
+    if lhs is None or rhs is None:
+        return None
+    left = typecast_time(lhs)
+    right = typecast_time(rhs)
+    return (
+        (left.hour * 60 * 60 * 1000000) +
+        (left.minute * 60 * 1000000) +
+        (left.second * 1000000) +
+        (left.microsecond) -
+        (right.hour * 60 * 60 * 1000000) -
+        (right.minute * 60 * 1000000) -
+        (right.second * 1000000) -
+        (right.microsecond)
+    )
+
+
+def _sqlite_timestamp_diff(lhs, rhs):
+    if lhs is None or rhs is None:
+        return None
+    left = typecast_timestamp(lhs)
+    right = typecast_timestamp(rhs)
+    return duration_microseconds(left - right)
+
+
+def _sqlite_regexp(re_pattern, re_string):
+    if re_pattern is None or re_string is None:
+        return None
+    return bool(re_search(re_pattern, str(re_string)))
+
+
+def _sqlite_acos(x):
+    if x is None:
+        return None
+    return acos(x)
+
+
+def _sqlite_asin(x):
+    if x is None:
+        return None
+    return asin(x)
+
+
+def _sqlite_atan(x):
+    if x is None:
+        return None
+    return atan(x)
+
+
+def _sqlite_atan2(y, x):
+    if y is None or x is None:
+        return None
+    return atan2(y, x)
+
+
+def _sqlite_bitxor(a, b):
+    if a is None or b is None:
+        return None
+    return a ^ b
+
+
+def _sqlite_ceiling(x):
+    if x is None:
+        return None
+    return ceil(x)
+
+
+def _sqlite_cos(x):
+    if x is None:
+        return None
+    return cos(x)
+
+
+def _sqlite_cot(x):
+    if x is None:
+        return None
+    return 1 / tan(x)
+
+
+def _sqlite_degrees(x):
+    if x is None:
+        return None
+    return degrees(x)
+
+
+def _sqlite_exp(x):
+    if x is None:
+        return None
+    return exp(x)
+
+
+def _sqlite_floor(x):
+    if x is None:
+        return None
+    return floor(x)
+
+
+def _sqlite_ln(x):
+    if x is None:
+        return None
+    return log(x)
+
+
+def _sqlite_log(x, base):
+    if x is None or base is None:
+        return None
+    # Arguments reversed to match SQL standard.
+    return log(base, x)
+
+
+def _sqlite_lpad(text, length, fill_text):
+    if text is None or length is None or fill_text is None:
+        return None
+    if len(text) >= length:
+        return text[:length]
+    return (fill_text * length)[:length - len(text)] + text
```

Comment:
Afaict we can just use `str.rjust` here.

Reply:
Ooh. Why didn't I think of that. Depends on whether `str.rjust()` supports multiple characters for the fill text. I only see `fillchar` in the docs, but we could still provide a fast-path if `fill_text` has a single character…

In addition we could consider `str.ljust()` for `_sqlite_lpad()` if it's viable.

Another question is whether SQL `LPAD` and `RPAD` should support multiple characters for the fill. It's this just a quirk of our implementation? We should check if we have tests with multiple fill characters either way to avoid regression.

Reply:
It looks like other databases support multiple characters for the fill, while indeed rjust/ljust don't. I think we should stick with a single implementation rather than try have a fast path.

Multiple character tests already exist: https://github.com/django/django/blob/40165eecc40f9e223702a41a0cb0958515bb1f82/tests/db_functions/text/test_pad.py

Reply:
I quickly compared your new implementation and it came out faster in all tests except where the length was shorter than the string. I think this is fine since it should be rare that padding is used to crop.

```
In [4]: %timeit _sqlite_lpad("barbecue", 8, " ")
224 ns ± 21.8 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [5]: %timeit _sqlite_lpad_2("barbecue", 8, " ")
179 ns ± 6.21 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [6]: %timeit _sqlite_lpad("barbecue", 2, " ")
248 ns ± 1.47 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [7]: %timeit _sqlite_lpad_2("barbecue", 2, " ")
288 ns ± 3.44 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [8]: %timeit _sqlite_lpad("barbecue", 20, " ")
414 ns ± 2.24 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [9]: %timeit _sqlite_lpad_2("barbecue", 20, " ")
388 ns ± 3.53 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

Reply:
> I think this isn't a great trade-off to make, the special branch for "is the length exactlyt the same" will cost more than it saves across a range of different input string types. Only a few will match it.

Yeah, I wondered whether the exact length check might be detrimental.

> Regardless this is a rarely used DB function, this is too much time spent optimizing it already :)

😄 Yup. Happy with your solution.